utils/pattern/src/parser/mod.rs

```diff
@@ -0,0 +1,374 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+pub mod error;
+
+use crate::{pattern::PatternIterator, token::PatternToken};
+pub use error::ParserError;
+use std::{fmt::Debug, str::FromStr};
+
+#[derive(PartialEq)]
+enum ParserState {
+    Default,
+    Placeholder,
+    QuotedLiteral,
+    Apostrophe { quoted: bool },
+}
+
+impl Default for ParserState {
+    fn default() -> Self {
+        Self::Default
+    }
+}
+
+macro_rules! advance_state {
```

Comment:
Why is this not an &mut self method?

Reply:
because it would increase parser instruction count by 3.7%:

```
iai_parse
  Instructions:                7094 (+3.758959%)
  L1 Accesses:                 9680 (+5.584642%)
  L2 Accesses:                    4 (  +inf%)
  RAM Accesses:                  67 (+1.515152%)
  Estimated Cycles:           12045 (+4.939885%)
```

Reply:
hmm, can't repro! Switching!